rt_gene/src/rt_gene/gaze_estimation_models_pytorch.py

```diff
@@ -0,0 +1,401 @@
+        self.left_features = nn.Sequential(*_left_features)
+
+        _right_features = [module for module in _right_model.features.children()]
+        self.right_features = nn.Sequential(
+            *list(_right_model.features.children())
+        )
+
+        for param in self.left_features.parameters():
+            param.requires_grad = True
+        for param in self.right_features.parameters():
+            param.requires_grad = True
+
+        _num_ftrs = _left_model.classifier[1].in_features + _right_model.classifier[1].in_features + 2  # left, right and head_pose
+        self.classifier = GazeEstimationAbstractModel._create_fc_layers(num_out=num_out, num_features=_num_ftrs)
+        GazeEstimationAbstractModel._init_weights(self.modules())
+
+    def forward(self, left_eye, right_eye, headpose):
+        left_x = self.left_features(left_eye)
+        left_x = left_x.mean([2, 3])
+
+        right_x = self.right_features(right_eye)
+        right_x = right_x.mean([2, 3])
+
+        concat = torch.cat((left_x, right_x, headpose), dim=1)
+
+        fc_output = self.classifier(concat)
+
+        # angular_output = fc_output[:, :2]
+        #
+        # sigma = fc_output[:, 2:3]
+        # sigma = sigma.view(-1, 1).expand(sigma.size(0), 2)
+
+        return fc_output
+
+
+class GazeEstimationModelVGG(GazeEstimationAbstractModel):
+
+    def __init__(self, num_out=2):  # phi, theta
+        super(GazeEstimationModelVGG, self).__init__()
+        _left_model = models.vgg16_bn(pretrained=True)
+        _right_model = models.vgg16_bn(pretrained=True)
+
+        # remove the last ConvBRelu layer
+        _left_modules = [module for module in _left_model.features]
+        _left_modules.append(_left_model.avgpool)
+        self.left_features = nn.Sequential(*_left_modules)
+
+        _right_modules = [module for module in _right_model.features]
+        _right_modules.append(_right_model.avgpool)
+        self.right_features = nn.Sequential(*_right_modules)
+
+        for param in self.left_features.parameters():
+            param.requires_grad = True
+        for param in self.right_features.parameters():
+            param.requires_grad = True
+
+        _num_ftrs = _left_model.classifier[0].in_features + _right_model.classifier[0].in_features + 2  # left, right and head_pose
+        self.classifier = GazeEstimationAbstractModel._create_fc_layers(num_out=num_out, num_features=_num_ftrs)
+        GazeEstimationAbstractModel._init_weights(self.modules())
+
+    def forward(self, left_eye, right_eye, head_pose):
+        left_x = self.left_features(left_eye)
+        left_x = torch.flatten(left_x, 1)
+
+        right_x = self.right_features(right_eye)
+        right_x = torch.flatten(right_x, 1)
+
+        concat = torch.cat((left_x, right_x, head_pose), dim=1)
+
+        fc_output = self.classifier(concat)
+
+        return fc_output
+
+
+if __name__ == "__main__":
```

Comment:
This looks like testing code which we can get rid off.

Reply:
agreed